src/oscar/navigation_menu_registry.py

```diff
@@ -0,0 +1,237 @@
+        }
+
+
+_dashboard_navigation_menu = defaultdict(Menu)
+_marked_for_deregistration = set()
+
+
+def get_legacy_parent_menu(lookup_label_or_index):
+    """Should help ease migration process from menus declared through `settings.OSCAR_DASHBOARD_NAVIGATION`"""
+    is_lookup_by_position = isinstance(lookup_label_or_index, int)
+    if is_lookup_by_position:
+        return settings.OSCAR_DASHBOARD_NAVIGATION[lookup_label_or_index]
+
+    # parent menu is to be identified by label
+    for index, parent_menu in enumerate(settings.OSCAR_DASHBOARD_NAVIGATION):
+        if parent_menu["label"] == lookup_label_or_index:
+            return parent_menu
+    raise ValueError(
+        f"Parent with {'position' if is_lookup_by_position else 'label'} '{lookup_label_or_index}' not found"
+    )
+
+
+def register_dashboard_menu(menu_or_function=None, parent_id=None):
+    if menu_or_function is None:
+        def decorator(decorated_function):
+            register_dashboard_menu(decorated_function, parent_id)
+            return decorated_function
+
+        return decorator
+
+    menu = menu_or_function() if callable(menu_or_function) else menu_or_function
+    if menu is None:
+        return
+
+    if isinstance(menu, dict):
+        menu = Menu.from_dict(menu)
+
+    if parent_id is None:
+        try:
+            _dashboard_navigation_menu[menu.identifier] += menu
+        except TypeError:
+            # `defaultdict` tried to instantiate a default Menu object but could not because
+            # 'label' positional argument is required
+            menu.auto_position_if_applicable(len(_dashboard_navigation_menu))
+            _dashboard_navigation_menu[menu.identifier] = menu
+    else:
+        # Placeholder instance will be used temporarily if child (menu) was registered ahead of
+        # its parent. It will be replaced (merged) into the actual parent once it's registered.
+        _dashboard_navigation_menu.setdefault(parent_id, Menu.placeholder(parent_id)).add_child(
+            menu).auto_position_if_applicable(len(_dashboard_navigation_menu))
+
+
+def unregister_dashboard_menu(parent_id, child_id=None):
+    _marked_for_deregistration.add((parent_id, child_id))
+
+
+_search_for_dashboard_navigation_menu = False
+
+
+def get_dashboard_navigation_menu():
+    global _search_for_dashboard_navigation_menu
+    if not _search_for_dashboard_navigation_menu:
+        list(_get_app_submodules("navigation_menu"))
+        _search_for_dashboard_navigation_menu = True
+
+    for parent_id, child_id in _marked_for_deregistration:
+        if child_id is not None:
+            parent_menu = _dashboard_navigation_menu.get(parent_id)
+            if parent_menu:
+                _dashboard_navigation_menu[parent_id] = parent_menu.remove_child(child_id)
+            else:
+                raise ValueError(f"Parent menu with the ID '{parent_id}' does not exist")
+        else:
+            del _dashboard_navigation_menu[parent_id]
```

Comment:
Please put the code that deals with removing deregistered menus in a separate function.